src/Maestro/FeedCleaner/FeedCleaner.cs

```diff
@@ -0,0 +1,369 @@
+        public FeedCleaner(
+            ILogger<FeedCleaner> logger,
+            BuildAssetRegistryContext context,
+            IOptions<FeedCleanerOptions> options)
+        {
+            Logger = logger;
+            Context = context;
+            _httpClient = new HttpClient();
+            _options = options;
+            AzureDevOpsClients = new Dictionary<string, IAzureDevOpsClient>();
+            foreach (string account in _options.Value.AzdoAccounts)
+            {
+                AzureDevOpsClients.Add(account, GetAzureDevOpsClientForAccountAsync(account));
+            }
+        }
+
+        public ILogger<FeedCleaner> Logger { get; }
+        public BuildAssetRegistryContext Context { get; }
+
+        public FeedCleanerOptions Options => _options.Value;
+
+        public Dictionary<string, IAzureDevOpsClient> AzureDevOpsClients { get; set; }
+
+        private readonly HttpClient _httpClient;
+        private readonly IOptions<FeedCleanerOptions> _options;
+
+        public Task<TimeSpan> RunAsync(CancellationToken cancellationToken)
+        {
+            return Task.FromResult(TimeSpan.FromMinutes(5));
+        }
+
+        /// <summary>
+        /// Updates assets that are now available from one of the non-stable feeds,
+        /// delete those package versions from the stable feeds and delete any feeds 
+        /// where all packages versions have been deleted every day at 2 AM.
+        /// </summary>
+        [CronSchedule("0 0 2 1/1 * ? *", TimeZones.PST)]
+        public async Task CleanManagedFeedsAsync()
+        {
+            if (Options.Enabled)
+            {
+                Dictionary<string, Dictionary<string, HashSet<string>>> packagesInReleaseFeeds =
+                    await GetPackagesForReleaseFeedsAsync();
+
+                foreach (var azdoAccount in Options.AzdoAccounts)
+                {
+                    var azdoClient = AzureDevOpsClients[azdoAccount];
+                    List<AzureDevOpsFeed> allFeeds = await azdoClient.GetFeedsAsync(azdoAccount);
+                    IEnumerable<AzureDevOpsFeed> managedFeeds = allFeeds.Where(f => Regex.IsMatch(f.Name, MaestroManagedFeedPattern));
+
+                    foreach (var feed in managedFeeds)
+                    {
+                        await PopulatePackagesForFeedAsync(feed);
+                        foreach (var package in feed.Packages)
+                        {
+                            HashSet<string> updatedVersions =
+                                await UpdateReleasedVersionsForPackageAsync(feed, package, packagesInReleaseFeeds);
+
+                            await DeletePackageVersionsFromFeedAsync(feed, package.Name, updatedVersions);
+                        }
+
+                        // We may have deleted all packages in the previous operation, if so, we should delete the feed,
+                        // refresh the packages in the feed to check this.
+                        await PopulatePackagesForFeedAsync(feed);
```

Comment:
It looks like this could throw. Does this need to be handled here? Otherwise, I think...the exception would propagate outward and mean that we couldn't clean any other feeds?

Ideally if you can't clean packages from a feed, or delete a feed, then you could just continue onto the next package.

Reply:
You're right, we should just move on in error cases. Fixed.